Lib/sysconfig.py

```diff
@@ -212,13 +212,38 @@ def _expand_vars(scheme, vars):
     return res
 
 
-def _get_default_scheme():
+def get_default_scheme():
     if os.name == 'posix':
         # the default scheme for posix is posix_prefix
         return 'posix_prefix'
     return os.name
 
 
+def _get_preferred_schemes():
+    if os.name == 'nt':
+        return {
+            'prefix': 'nt',
+            'home': 'posix_home',
+            'user': 'nt_user',
+        }
+    if sys.platform == 'darwin' and sys._framework:
+        return {
+            'prefix': 'posix_prefix',
+            'home': 'posix_home',
+            'user': 'osx_framework_user',
+        }
+    return {
+        'prefix': 'posix_prefix',
+        'home': 'posix_home',
+        'user': 'posix_user',
+    }
+
+
+def get_preferred_scheme(key):
+    scheme = _get_preferred_schemes()[key]
+    if scheme not in _INSTALL_SCHEMES:
+        raise KeyError(key)
```

Comment:
The main intention is to guard against incorrect implementation, since `_get_preferred_schemes()` is expected to be modified by redistributors and implementers. Say an implementer makes a typo:

```python
def _get_preferred_schemes():
    return {
        "prefix": "posix_prefix",
        "home": "posix_home",
        "user": "posix_uesr",  # Oops!
    }
```

And when the value is later used:

```python
scheme = sysconfig.get_preferred_scheme("user")

...  # Much later...

paths = sysconfig.get_paths(scheme)
```

This check would catch that `posix_uesr` is not actually a valid scheme and error early, instead of later when the user actually tries to use the (invalid) scheme.

Reply:
Can we raise a ValueError with a better message then? "Key '{}' returned scheme '{}' which is not valid on this platform" (or something more correct)

Reply:
I couldn’t think of a better message and implemented yours for now, any ideas are very welcomed. Ideally I want to point users to the implementer/redistributor instead of reporting the error to CPython, but that’s probably not really feasible.